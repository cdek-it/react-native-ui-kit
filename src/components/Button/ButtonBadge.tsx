import { memo, useCallback, useMemo, useState } from 'react'
import {
  type LayoutChangeEvent,
  type LayoutRectangle,
  View,
  type ViewStyle,
} from 'react-native'

import { Badge } from '../Badge'

import { BaseButton } from './BaseButton'
import { useBasicButtonStyles } from './styles'
import type { ButtonBadgeProps, ButtonProps } from './types'

/**
 * Button component with badge
 * @param size - button size
 * @param shape - button shape
 * @param loading - button loading state
 * @param variant - button variant
 * @param disabled - button disabled state
 * @param iconOnly - button with only Icon
 * @param iconPosition - icon position
 * @param Icon - Tabler icon
 * @param label - button label
 * @param style - external style control for component
 * @param badgeSeverity - badge component severity level
 * @param badgeLabel - text label inside badge
 * @see BaseButton
 */
export const ButtonBadge = memo<ButtonProps & ButtonBadgeProps>(
  ({ badgeLabel, badgeSeverity, testID, ...props }) => {
    const buttonStyles = useBasicButtonStyles()
    const [badgeLayout, setBadgeLayout] = useState<LayoutRectangle>()

    const badgeContainerStyle = useMemo<ViewStyle>(
      () => ({
        position: 'absolute',
        top: badgeLayout ? -Math.round(badgeLayout.height / 2) : 0,
        right: badgeLayout ? -Math.round(badgeLayout.width / 2) : 0,
      }),
      [badgeLayout]
    )

    const onLayout = useCallback(
      (e: LayoutChangeEvent) => setBadgeLayout(e.nativeEvent.layout),
      []
    )

    return (
      <View>
        <BaseButton {...props} {...buttonStyles} />

        <Badge
          severity={badgeSeverity}
          style={badgeContainerStyle}
          testID={testID || 'buttonBadge'}
          onLayout={onLayout}
        >
          {badgeLabel}
        </Badge>
      </View>
    )
  }
)
