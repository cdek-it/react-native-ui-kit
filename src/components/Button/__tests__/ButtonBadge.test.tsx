import { render } from '@testing-library/react-native'

import { ButtonBadge } from '../ButtonBadge'

describe('ButtonSeverity component tests', () => {
  test('badge - basic', () => {
    const renderedButton = render(
      <ButtonBadge badgeLabel='Badge' badgeSeverity='danger' label='Button' />
    )

    expect(renderedButton.toJSON()).toMatchSnapshot()
  })
})
