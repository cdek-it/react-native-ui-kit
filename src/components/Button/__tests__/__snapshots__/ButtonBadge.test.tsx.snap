// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSeverity component tests badge - basic 1`] = `
<View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "borderColor": "rgba(255, 255, 255, 0.0001)",
        "borderRadius": 7,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 7,
        "height": 35,
        "justifyContent": "center",
        "maxHeight": 35,
        "minHeight": 35,
        "paddingHorizontal": 14,
        "paddingVertical": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 14,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
      testID="Button_Text"
    >
      Button
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "alignItems": "flex-start",
        },
        {
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
    testID="buttonBadge"
  >
    <View
      style={
        [
          {
            "borderRadius": 100,
            "height": 21,
            "justifyContent": "center",
            "paddingHorizontal": 7,
            "paddingVertical": 0,
          },
          {
            "backgroundColor": "#e26e63",
          },
        ]
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 10.5,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        Badge
      </Text>
    </View>
  </View>
</View>
`;
